modules/page-memory/src/main/java/org/apache/ignite/internal/pagememory/persistence/checkpoint/CheckpointReadWriteLockMetricSource.java

```diff
@@ -0,0 +1,136 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one or more
+ * contributor license agreements. See the NOTICE file distributed with
+ * this work for additional information regarding copyright ownership.
+ * The ASF licenses this file to You under the Apache License, Version 2.0
+ * (the "License"); you may not use this file except in compliance with
+ * the License. You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.ignite.internal.pagememory.persistence.checkpoint;
+
+import java.util.List;
+import java.util.concurrent.atomic.AtomicInteger;
+import java.util.function.IntSupplier;
+import org.apache.ignite.internal.metrics.AbstractMetricSource;
+import org.apache.ignite.internal.metrics.DistributionMetric;
+import org.apache.ignite.internal.metrics.IntGauge;
+import org.apache.ignite.internal.metrics.Metric;
+
+/**
+ * Metric source for checkpoint read lock operations.
+ *
+ * <p>This metric source tracks performance and contention characteristics of checkpoint read locks
+ * acquired by normal operations during database operation. These locks coordinate with the checkpoint
+ * write lock to ensure consistency.
+ */
+public class CheckpointReadWriteLockMetricSource extends AbstractMetricSource<CheckpointReadWriteLockMetricSource.Holder> {
+    private static final long[] LOCK_ACQUISITION_MILLIS = {
+            1,      // 1ms   - uncontended, fast path
+            10,     // 10ms  - minor contention
+            50,     // 100ms - moderate contention
+            100,    // 100ms - high contention
+            500,    // 500ms - checkpoint in progress?
+            1_000,  // 1s    - severe contention, reported as warning in logs
+            10_000  // 10s   - pathological case, shall be treated as an emergency error
+    };
+
+    private static final long[] LOCK_HOLD_MILLIS = {
+            1,      // 1ms   - very fast operation
+            10,     // 10ms  - fast single-page operation
+            50,     // 50ms  - multi-page operation
+            100,    // 100ms - complex operation
+            500,    // 500ms - batch operation
+            1_000,  // 1s    - too large batch or slow I/O
+            10_000  // 10s  - pathologically large operation
+    };
+
+    /** Counter for threads currently waiting for checkpoint read lock. */
+    private final AtomicInteger readLockWaitingThreadsGauge = new AtomicInteger(0);
```

Comment:
Let's make it a `LongAdder` just in case. We can even make it `LongAdderMetric` right away